lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -67,13 +71,33 @@ newtype Transaction a
 
 -- | Run a transaction on the given connection.
 runTransaction :: MonadIO m => Connection -> Transaction a -> m a
-runTransaction conn@(Connection _ _ conn0) (Transaction f) =
-  -- TODO some sensible retry logic
-  liftIO do
-    Sqlite.execute_ conn0 "BEGIN"
-    result <- f conn
-    Sqlite.execute_ conn0 "COMMIT"
+runTransaction conn (Transaction f) = liftIO do
+  uninterruptibleMask \restore -> do
```

Comment:
Is there no good way we can rephrase this with `bracket`? 
It's well known in the Haskell community that using "masks" is extremely difficult to get right, and while I believe you've given it sufficient thought here I think it imposes a burden on all future changes to this code.
The fact that it wraps retry logic makes it even more complex.

I'd much much prefer we use some combination of `bracket`, `onException`, `finally`, and friends instead.

Note that the `unliftio` variants of these helpers use `uninterruptible` masking on their cleanup blocks. I'm not 100% of the implications of not being uninterruptible on the setup block, but I suspect these would be suitable for our use-case here.